ENH:  Calculate WCS pixel area and scales as Quantity

pllim opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

This pull request is to implement new `WCS` methods that return pixel scales and area as `Quantity`.

I am not convinced that the low-level functions in `utils` should be deprecated. If people like the non-Quantity stuff, they are free to use them still.

I am also not convinced that we should force the outputs to be in arcsec. If `cunit` says `'deg'`, I think it is reasonable to return degrees. People can easily convert to arcsec at their own leisure.

I also took the liberty to fix some PEP 8 stuff, so I know if I accidentally introduced new violation or not (I try not to). In that process, I also found a duplicate test function, which I deleted here.

If you disagree, we can discuss during the review process.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #6523

pllim commented on astropy/wcs/wcs.py:
```diff
@@ -886,7 +891,6 @@ def _write_det2im(self, hdulist):
             return
         dist = 'D2IMDIS'
         d_kw = 'D2IM'
-        err_kw = 'D2IMERR'
```
Note: Not used.

pllim commented on astropy/wcs/wcs.py:
```diff
@@ -1008,10 +1012,10 @@ def _write_distortion_kw(self, hdulist, dist='CPDIS'):
         if dist == 'CPDIS':
             d_kw = 'DP'
-            err_kw = 'CPERR'
+            # err_kw = 'CPERR'
```
Also not used. But somehow here I commented instead of deleting. I should be consistent, but I want to hear what others say first.

mhvk replied:
Just delete it!

astrofrog commented on astropy/wcs/wcs.py:
```diff
@@ -2446,6 +2450,26 @@ def sip_foc2pix(self, *args):
+        """
+        from astropy.wcs import utils as wcs_utils  # Avoid circular import
+        val = wcs_utils.proj_plane_pixel_area(self)
+        unit = u.Unit(self.wcs.cunit[0]) * u.Unit(self.wcs.cunit[1])  # 2D only
```
This should probably error if not 2D?

pllim replied:
`astropy.wcs.utils.proj_plane_pixel_area` already checks it.

pllim replied:
Is this check insufficient?

https://github.com/astropy/astropy/blob/4eef52a1c95471b837949c5a8f862a01e4ca1bfa/astropy/wcs/utils.py#L390-L392

If we need to add more check, I feel like it has to be added inside the `utils` function, not the high-level method wrapper.

astrofrog reviewed (dismissed):
I remember why these were separate functions - it is because for the area at least I think this was mainly intended for celestial WCS. We should test this with non-celestial WCS to see if it produces nonsensical results. If so, then we might want to raise an error if the WCS is not celestial?

pllim commented:
I added a test, but I am no WCS expert, so if there are other cases to consider, please let me know.

mhvk commented on astropy/wcs/wcs.py:
```diff
@@ -2446,6 +2450,26 @@ def sip_foc2pix(self, *args):
+        from astropy.wcs import utils as wcs_utils  # Avoid circular import
+        val = wcs_utils.proj_plane_pixel_scales(self)
+        units = [u.Unit(x) for x in self.wcs.cunit]
+        return list(map(operator.mul, val, units))  # Can have different units
```
I know it is the same thing, but to me the following is clearer:
```
return [value * unit for (v, unit) in zip(values, units)]
```

mhvk commented on astropy/wcs/wcs.py:
```diff
@@ -2446,6 +2450,26 @@ def sip_foc2pix(self, *args):
+
+        """
+        from astropy.wcs import utils as wcs_utils  # Avoid circular import
+        val = wcs_utils.proj_plane_pixel_area(self)
```
Maybe use `value`? (total nitpick)

mhvk commented on astropy/wcs/wcs.py:
```diff
@@ -2446,6 +2450,26 @@ def sip_foc2pix(self, *args):
+
+        """
+        from astropy.wcs import utils as wcs_utils  # Avoid circular import
+        val = wcs_utils.proj_plane_pixel_scales(self)
```
Maybe use `values`? (More total nitpicking...)

pllim replied:
Done. I thought I was pretty progressive in using 3 chars, but apparently not progressive enough. 😸

mhvk reviewed (commented):
This looks good to me! Agreed with @astrofrog about the pixel area, though I think the code as written should give sensible results (even if the units are "Å arcsec" or so...). So, left are only the most picky of nits!

And agree very much not to enforce a particular unit for the output but rather assume the writers of the file knew what the preferred unit was.

pllim commented:
p.s. Failures are unrelated. Will need to rebase after #10874 is merged.

pllim commented:
Update: Rebased and tests passed. Please let me know if I overlooked any comments. Thanks for the reviews!

mcara commented on astropy/wcs/wcs.py:
```diff
@@ -2446,6 +2448,26 @@ def sip_foc2pix(self, *args):
+        but returns a list of `~astropy.units.Quantity`.
+
+        """
+        from astropy.wcs import utils as wcs_utils  # Avoid circular import
```
Why not move the code from `utils.py` here as `_ proj_plane_pixel_scales()` and then call it from both functions? Doing this, I think, would avoid the need for this import within the method.

mcara replied:
@pllim I did not do any timing tests on this specific PR, but I do not think the answer from https://stackoverflow.com/a/32151422 is accurate. The benchmark does not compare timings of global import vs local import.

I did a quick test on a function that uses numpy's `sin()` to do a simple computation:

#### Local import
```python
In [1]: def f(x):
   ...:     from numpy import sin
   ...:     return sin(x)
   ...:

In [2]: %timeit f(0.5)
1.08 µs ± 10.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: exit
```

#### Global import
```
In [1]: from numpy import sin

In [2]: def f(x):
   ...:     return sin(x)
   ...:

In [3]: %timeit f(0.5)
587 ns ± 7.57 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

pllim replied:
Let's use a more realistic example:

```python
>>> from astropy.io import fits
>>> from astropy.utils.data import get_pkg_data_filename
>>> from astropy.wcs import WCS

>>> path = get_pkg_data_filename("data/dist_lookup.fits.gz", package='astropy.wcs.tests')
>>> with fits.open(path) as hdulist:
...     w = WCS(hdulist[1].header, hdulist)

>>> # Time this first before wcs.utils import below.
>>> # This was done in IPython. I just have a habit to use >>> in markdown.
>>> %timeit w.proj_plane_pixel_scales()
87.6 µs ± 78.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

>>> from astropy import units as u
>>> from astropy.wcs.utils import proj_plane_pixel_scales
>>> # Same calculation as the class method
>>> def proj_this(w):
...     values = proj_plane_pixel_scales(w)
...     units = [u.Unit(x) for x in w.wcs.cunit]
...     return [value * unit for (value, unit) in zip(values, units)]

>>> # Time the standalone function
>>> # This was done in IPython. I just have a habit to use >>> in markdown.
>>> %timeit proj_this(w)
87.1 µs ± 149 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Okay, that's 0.5 microseconds improvement in something that takes 87 microseconds to calculate. Is this small improvement worth the extra code complexity in moving the `utils` functions completely out of `utils` just to avoid circular import while being able to do module level import?

mcara replied:
I do not think the code is more complex, it is easier to read (follow), and it is faster. I suspect that is the reason why my older phone is "getting slower" :)

mcara replied:
Could you, please, for completeness, benchmark `utils.proj_plane_pixel_scales(w)`. I am getting like 1/3 of your timing.

pllim replied:
What's the point of that? It doesn't apply units, so it is like comparing orange to apple.

mcara commented on astropy/wcs/wcs.py:
```diff
@@ -2446,6 +2448,26 @@ def sip_foc2pix(self, *args):
+        but returns a `~astropy.units.Quantity`.
+
+        """
+        from astropy.wcs import utils as wcs_utils  # Avoid circular import
```
Same as in https://github.com/astropy/astropy/pull/10872/files#r506769107

mcara commented on astropy/wcs/wcs.py:
```diff
@@ -2446,6 +2448,26 @@ def sip_foc2pix(self, *args):
                    docstrings.RETURNS('pixel coordinates', 8))
 
+    def proj_plane_pixel_scales(self):
+        """Same as :func:`astropy.wcs.utils.proj_plane_pixel_scales`
+        but returns a list of `~astropy.units.Quantity`.
+
+        """
```
If this function is going to be the new "workhorse", and also because it is a method of the main class `WCS` from this sub-package, I think it would be better to simply copy the docstring from the `utils` function and maybe make a mention like "For unit-un-aware version - see ..." (something like that)?

mhvk replied:
:+1: - and one can have a `See Also` section to refer the the unit unaware function.

mhvk commented on astropy/wcs/wcs.py:
```diff
@@ -33,6 +33,7 @@
 import io
 import itertools
 import os
+import operator
```
I think you may not need `operator` any more since you followed my recommendation!

pllim replied:
Ops, good catch!

pllim commented on astropy/wcs/wcs.py:
```diff
@@ -2449,22 +2448,83 @@ def sip_foc2pix(self, *args):
+
+        """  # noqa: E501
+        from astropy.wcs.utils import proj_plane_pixel_scales  # Avoid circular import
+        values = proj_plane_pixel_scales(self.celestial)
```
Re-reading the docstring again, I think this is more correct? That is, by passing in `self.celestial` instead of `self`?

mcara replied:
There is nothing conceptually wrong with passing the "entire" `self`. Let's say `self` has 2 celestial axes and one spectral. This function would return something like `CDELT` for each axis. It should be fine. I think you should pass the whole WCS object.

pllim replied:
Ah, okay, I'll change it back then. This snippet from the `utils` docstring confused me.

```rst
    .. note::
        In order to compute the scales corresponding to celestial axes only,
        make sure that the input `~astropy.wcs.WCS` object contains
        celestial axes only, e.g., by passing in the
        `~astropy.wcs.WCS.celestial` WCS object.
```

mcara replied:
That is just to direct the reader what to do if that is what they want (i.e., to get the scale _only_ for celestial axes and exclude non-celestial axes)

mcara replied:
For the "area", it is highly unlikely that anyone would want to compute the area of 1 celestial axis and 1 spectral axis, even though mathematically one can do anything they want. Or, I am not sure most people would find it useful to compute the "volume" of a 2D celestial axes X 1D spectral. Probably, in most cases, user would want to separate these kind of axes but I may be wrong. If there is demand to compute volumes, let me know.

I personally find `proj_plane_pixel_scales` less useful than `proj_plane_pixel_area` because the later is used to approximate "mean pixel scale" (**geometric mean**) which, I personally find more useful than knowing scale along each axes. It also can be used to equalize pixel intensity when displaying an image (like it is done in FLT images).

pllim replied:
I did use `proj_plane_pixel_scales` when I was scratching my head over spacetelescope/stginga#205 😸

pllim commented:
I think I have addressed all the comments thus far, except for the suggestion for moving out stuff out of `utils.py`, as I am still not convinced it is worth the effort if the little extra local import is the only thing causing it to happen.

pllim commented:
@astrofrog , do you still have objection to this PR?

pllim commented:
I'll take silence as "no objection".
